File: sdk/src/main/java/com/cxense/cxensesdk/db/DatabaseHelper.java
package com.cxense.cxensesdk.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.List;

/**
 * Class for working with database
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */

public class DatabaseHelper extends SQLiteOpenHelper {
    static final int DATABASE_VERSION = 1;
    private static final String DATABASE_NAME = "tracks.db";

    public DatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    static void parseCursor(Cursor cursor, ContentValues values) {
        String[] columns = cursor.getColumnNames();
        int length = columns.length;
        for (int i = 0; i < length; i++) {
            switch (cursor.getType(i)) {
                case Cursor.FIELD_TYPE_NULL:
                    values.putNull(columns[i]);
                    break;
                case Cursor.FIELD_TYPE_INTEGER:
                    values.put(columns[i], cursor.getLong(i));
                    break;
                case Cursor.FIELD_TYPE_FLOAT:
                    values.put(columns[i], cursor.getDouble(i));
                    break;
                case Cursor.FIELD_TYPE_STRING:
                    values.put(columns[i], cursor.getString(i));
                    break;
                case Cursor.FIELD_TYPE_BLOB:
                    values.put(columns[i], cursor.getBlob(i));
                    break;
            }
        }
    }

    static ContentValues cursorRowToContentValues(Cursor cursor) {
        ContentValues values = new ContentValues();
        parseCursor(cursor, values);
        return values;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        for (int i = 0; i < DATABASE_VERSION; i++)
            migrate(db, i);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        for (int i = oldVersion; i < newVersion; i++)
            migrate(db, i);
    }

    public long save(DatabaseObject databaseObject) {
        ContentValues values = databaseObject.toContentValues();
        SQLiteDatabase db = getWritableDatabase();
        Integer id = databaseObject.id;
        String tableName = databaseObject.getTableName();
        if (id == null)
            return db.insert(tableName, null, values);
        return db.update(tableName, values, DatabaseObject._ID + "= ?", new String[]{"" + id});
    }

    public int delete(DatabaseObject databaseObject) {
        Integer id = databaseObject.id;
        if (id == null)
            return -1;
        return delete(databaseObject.getTableName(), DatabaseObject._ID + "= ?", new String[]{"" + id});
    }

    public int delete(String tableName, String whereClause, String[] whereArgs) {
        return getWritableDatabase().delete(tableName, whereClause, whereArgs);
    }

    public List<ContentValues> query(String tableName, String[] columns, String selection, String[] selectionArgs,
                                     String groupBy, String having, String orderBy) {
        return query(tableName, columns, selection, selectionArgs, groupBy, having, orderBy, null);
    }

    public List<ContentValues> query(String tableName, String[] columns, String selection, String[] selectionArgs,
                                     String groupBy, String having, String orderBy, String limit) {
        Cursor cursor = null;
        List<ContentValues> values = new ArrayList<>();
        try {
            cursor = getReadableDatabase().query(tableName, columns, selection, selectionArgs,
                    groupBy, having, orderBy, limit);
            while (cursor.moveToNext()) {
                values.add(cursorRowToContentValues(cursor));
            }
            return values;
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    /**
     * Migrates database from {@code oldVersion} to {@code oldVersion + 1}
     *
     * @param db         database
     * @param oldVersion current database version
     */
    protected void migrate(SQLiteDatabase db, int oldVersion) {
        switch (oldVersion) {
            case 0:
                // database does not exist, create it
                db.execSQL("CREATE TABLE " + EventRecord.TABLE_NAME + " ("
                        + EventRecord._ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
                        + EventRecord.EVENT_CUSTOM_ID + " TEXT,"
                        + EventRecord.EVENT + " TEXT,"
                        + EventRecord.TIME + " INTEGER,"
                        + EventRecord.CKP + " TEXT,"
                        + EventRecord.RND + " TEXT,"
                        + EventRecord.EVENT_TYPE + " TEXT,"
                        + EventRecord.SPENT_TIME + " INTEGER,"
                        + EventRecord.IS_SENT + " INTEGER);");
                break;
        }
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/db/DatabaseHelper.kt
package com.cxense.cxensesdk.db

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.provider.BaseColumns
import androidx.annotation.RestrictTo

@RestrictTo(RestrictTo.Scope.LIBRARY)
open class DatabaseHelper(
    context: Context
) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase?) {
        for (v in 0 until DATABASE_VERSION)
            migrate(db, v)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        for (v in oldVersion until newVersion)
            migrate(db, v)
    }

    fun migrate(db: SQLiteDatabase?, version: Int) {
        when (version) {
            0 -> // database does not exist, create it
                // database does not exist, create it
                db?.execSQL(
                    """
                        CREATE TABLE ${EventRecord.TABLE_NAME} (
                        ${BaseColumns._ID} INTEGER PRIMARY KEY AUTOINCREMENT,
                        ${EventRecord.CUSTOM_ID} TEXT,
                        ${EventRecord.DATA} TEXT,
                        ${EventRecord.TIME} INTEGER,
                        ${EventRecord.CKP} TEXT,
                        ${EventRecord.RND} TEXT,
                        ${EventRecord.TYPE} TEXT,
                        ${EventRecord.SPENT_TIME} INTEGER,
                        ${EventRecord.IS_SENT} INTEGER
                        );
                    """.trimIndent()
                )
        }
    }

    fun save(eventRecord: EventRecord): Long =
        eventRecord.id?.let { id ->
            writableDatabase.update(
                EventRecord.TABLE_NAME,
                eventRecord.toContentValues(),
                "${BaseColumns._ID} = ?",
                arrayOf(id.toString())
            ).toLong()
        } ?: writableDatabase.insert(
            EventRecord.TABLE_NAME,
            null,
            eventRecord.toContentValues()
        )

    fun delete(eventRecord: EventRecord): Int =
        eventRecord.id?.let { id ->
            delete("${BaseColumns._ID} = ?", id.toString())
        } ?: -1

    fun delete(whereClause: String?, vararg whereArgs: String): Int =
        writableDatabase.delete(EventRecord.TABLE_NAME, whereClause, whereArgs)

    fun query(
        columns: Array<String>? = EventRecord.COLUMNS,
        selection: String? = null,
        selectionArgs: Array<out String>? = null,
        groupBy: String? = null,
        having: String? = null,
        orderBy: String? = "${EventRecord.TIME} ASC",
        limit: String? = null
    ): List<ContentValues> =
        readableDatabase.query(
            EventRecord.TABLE_NAME,
            columns,
            selection,
            selectionArgs,
            groupBy,
            having,
            orderBy,
            limit
        ).use { c ->
            generateSequence { if (c.moveToNext()) c else null }
                .map(::cursorRowToContentValues)
                .toList()
        }

    companion object {
        const val DATABASE_VERSION = 1
        private const val DATABASE_NAME = "tracks.db"

        @JvmStatic
        private fun cursorRowToContentValues(c: Cursor): ContentValues =
            ContentValues().apply {
                for (i in 0 until c.columnCount) {
                    val name = c.columnNames[i]
                    when (c.getType(i)) {
                        Cursor.FIELD_TYPE_BLOB -> put(name, c.getBlob(i))
                        Cursor.FIELD_TYPE_FLOAT -> put(name, c.getFloat(i))
                        Cursor.FIELD_TYPE_INTEGER -> put(name, c.getLong(i))
                        else -> put(name, c.getString(i))
                    }
                }
            }
    }
}
